oauth2/src/main/java/com/linecorp/armeria/internal/common/auth/oauth2/AbstractOAuth2Request.java

```diff
@@ -18,207 +18,81 @@
-                throw onUnauthorizedError(response);
-        }
-        throw new UnsupportedResponseException(status, response.contentUtf8());
+        return HttpRequest.of(headers, HttpData.ofUtf8(bodyParams.toQueryString()));
     }
 
-    /**
-     * Composes {@link TokenRequestException} upon 400 Bad Request response
-     * as per <a href="https://datatracker.ietf.org/doc/html/rfc6749#section-5.2">[RFC6749], Section 5.2</a>.
-     * @param errorResponse response received from the server
-     * @return an instance of {@link TokenRequestException}
-     */
-    protected TokenRequestException onBadRequestError(AggregatedHttpResponse errorResponse) {
-        return TokenRequestException.parse(errorResponse.contentUtf8());
-    }
-
-    /**
-     * Composes {@link InvalidClientException} upon 401 Unauthorized response
-     * as per <a href="https://datatracker.ietf.org/doc/html/rfc6749#section-5.2">[RFC6749], Section 5.2</a> (invalid_client).
-     * @param errorResponse response received from the server
-     * @return an instance of {@link InvalidClientException}
-     */
-    protected TokenRequestException onUnauthorizedError(AggregatedHttpResponse errorResponse) {
-        final StringBuilder messageBuilder = new StringBuilder().append(errorResponse.status());
-        final String wwwAuthenticate = errorResponse.headers().get(HttpHeaderNames.WWW_AUTHENTICATE);
-        if (wwwAuthenticate != null) {
-            messageBuilder.append(": ").append(wwwAuthenticate);
-        }
-        final HttpData errorResponseContents = errorResponse.content();
-        if (!errorResponseContents.isEmpty()) {
-            messageBuilder.append(": ").append(errorResponseContents.toStringUtf8());
+    @Override
+    public final void addBodyParams(QueryParamsBuilder formBuilder) {
+        if (clientAuthentication != null) {
+            final String headerValue = clientAuthentication.asHeaderValue();
+            if (headerValue == null) {
+                clientAuthentication.addAsBodyParameters(formBuilder);
+            }
         }
-        return new InvalidClientException(messageBuilder.toString(), null);
+        addBodyParams0(formBuilder);
     }
 
-    /**
-     * Validates the content type of the response.
-     */
-    private static void validateContentType(AggregatedHttpResponse response, MediaType expectedType) {
-        final MediaType contentType = response.contentType();
-        if (contentType == null) {
-            // if omitted, assume that the type matches the expected
-            return;
-        }
-        final String mediaType = contentType.nameWithoutParameters();
-        if (!mediaType.equalsIgnoreCase(expectedType.nameWithoutParameters())) {
-            throw new UnsupportedMediaTypeException(mediaType,
-                                                    response.status().toString(), response.contentUtf8());
+    public abstract void addBodyParams0(QueryParamsBuilder formBuilder);
```

Comment:
Can we make this method `protected` and rename it to `doAddBodyParams()` sinces it's now part of the public API? `...0()` is the name we probably don't want to expose in the public API.

Reply:
This is an internal class so `addBodyParams0` is not exposed to the public API.
I wanted to reduce the mistake of omitting to call `super.addBodyParams()` when overriding a method by adding the `0`, `1`, and `2` suffixes.

That said, it is a good idea to prefix `do` to the internal methods.